### PR TITLE
ParallelTaskMapper replaced by swifty concurrentMap and concurrentCompactMap

### PR DESCRIPTION
**Pull request by adrien-coye**
### Abstract

ParallelTaskMapper has an API that complies with somthing we can inject in DI but this does not have a lot of sence IMO.

There is a case to match existing `map` and `compactMap` design on the `Collection` type.

ParallelTaskMapper was replaced by `concurrentMap` and `concurrentCompactMap`

### Details

- feat: modern API for concurrentMap and concurrentCompactMap.
- unit testsing with good coverage.
- improvements `concurrentMap` and `concurrentCompactMap` only rethrow.

BREAKING CHANGE: ParallelTaskMapper removed

### CODE REVIEW

File: Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                    try await accumulator.set(item: result, atIndex: index)
+                }
+            }
+
+            // await completion of all tasks.
+            try await taskGroup.waitForAll()
+        }
+
+        // Get the accumulated results.
+        let accumulated = await accumulator.compactAccumulation
```
Comment (PhilippeWeidmann) [inline Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift:62]: It looks like use `accumulator.compactAccumulation` in both `concurrentCompactMap` and `concurrentMap` is that an error ?

File: Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift
Comment (PhilippeWeidmann) [inline Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift:1]: A lot of code is shared in those to functions, can we refactor it in an internal function ?

File: Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift (hunk 1 above)
Comment (adrien-coye) [reply Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift:62]: It is not a bug as in both cases you want to return a collection of <Output> not a collection of <Output?>. (akin to what does the standard lib with `map` and `compactMap`)

The only difference been that a map() is expected to return the same number of arguments, where the compact map can return less items.

It all comes from the underlaying structure I'm using here an "accumulator". Not sure I can refactor given the constrains of concurency and type safety but I'll think about it.

File: Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift
Comment (adrien-coye) [reply Sources/InfomaniakCore/Asynchronous/Collection+concurrentMap.swift:1]: fixed

**Review by PhilippeWeidmann — APPROVED** [review]

File: Tests/InfomaniakCoreTests/Asynchronous/UTCollection+compactConcurrentMap.swift [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+
+        let result: [Int] = await collectionToProcess.concurrentCompactMap { item in
+            // We arbitrarily remove elements
+            if item % 10 == 0 {
+                return nil
+            }
+            return item * 10
+        }
+
+        // THEN
```
Comment (adrien-coye) [inline Tests/InfomaniakCoreTests/Asynchronous/UTCollection+compactConcurrentMap.swift:42]: incorect label

File: Tests/InfomaniakCoreTests/Asynchronous/UTCollection+compactConcurrentMap.swift [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+    private enum DomainError: Error {
+        case some
+    }
+
+    // MARK: - concurrentCompactMap
+
+    func testConcurrentCompactMapToArray() async {
+        // GIVEN
+        let collectionToProcess = Array(0 ... 50)
+
```
Comment (adrien-coye) [inline Tests/InfomaniakCoreTests/Asynchronous/UTCollection+compactConcurrentMap.swift:33]:
```suggestion

// WHEN
```

**Comment by adrien-coye** [comment]
Closing this PR, superseded by the open-source lib swift-concurrency https://github.com/Infomaniak/swift-concurrency

TODO, deprecate and eventually remove `ParallelTaskMapper` from this lib.

**Pull request closed without merging**